File: rubygems/pray-cli/lib/pray/publish_select.rb
# frozen_string_literal: true

module Pray
  module PublishSelect
    module_function

    PublishDestination = Struct.new(:name, :root, :server, :packages) do
      def initialize(name:, root: nil, server: nil, packages: [])
        super
      end
    end
    PublishCliDest = Struct.new(:to, :roots, :servers) do
      def initialize(to: [], roots: [], servers: [])
        super
      end
    end

    def resolve_publish_destinations(remotes, cli, project_root)
      if !cli.to.empty? && (!cli.roots.empty? || !cli.servers.empty?)
        raise Error.usage("publish --to cannot be combined with --root or --server")
      end
      return resolve_undeclared_destinations(cli) if remotes.empty?

      select_remotes(remotes, cli, project_root).map do |remote|
        PublishDestination.new(
          name: remote.name,
          root: remote.path && File.join(project_root, remote.path),
          server: remote.url,
          packages: remote.packages
        )
      end
    end

    def resolve_path_remote(remotes, to, root, project_root)
      path_remotes = remotes.select(&:path)
      if to
        remote = remotes.find { |entry| entry.name == to }
        raise Error.usage("unknown publish remote: #{to}") unless remote
        unless remote.path
          raise Error.usage("publish remote #{to} is a URL; yank, serve, and token need a path remote")
        end
        return File.join(project_root, remote.path)
      end
      if remotes.empty?
        raise Error.usage("requires --root PATH") unless root

        return root
      end
      if root
        matched = remotes.find { |remote| remote.path && root_matches?(root, remote.path, project_root) }
        raise Error.usage("path #{root} is not a declared publish remote") unless matched

        return root
      end
      if path_remotes.length == 1 && path_remotes.first.path
        return File.join(project_root, path_remotes.first.path)
      end

      raise Error.usage("say which path remote with --to NAME or --root PATH")
    end

    def resolve_undeclared_destinations(cli)
      if !cli.to.empty?
        raise Error.usage("Prayfile has no publish remotes; pass --root PATH or --server URL")
      end
      if cli.roots.empty? && cli.servers.empty?
        raise Error.unsupported("publish requires at least one --root PATH or --server URL")
      end

      dests = cli.roots.map { |root| PublishDestination.new(name: root, root: root, packages: []) }
      dests + cli.servers.map { |server| PublishDestination.new(name: server, server: server, packages: []) }
    end

    def select_remotes(remotes, cli, project_root)
      unless cli.to.empty?
        return cli.to.map do |name|
          remote = remotes.find { |entry| entry.name == name }
          raise Error.usage("unknown publish remote: #{name}") unless remote

          remote
        end
      end
      return remotes if cli.roots.empty? && cli.servers.empty?

      selected = cli.roots.map do |root|
        remote = remotes.find { |entry| entry.path && root_matches?(root, entry.path, project_root) }
        raise Error.usage("path #{root} is not a declared publish remote") unless remote

        remote
      end
      selected + cli.servers.map do |server|
        remote = remotes.find { |entry| entry.url == server }
        raise Error.usage("server #{server} is not a declared publish remote") unless remote

        remote
      end
    end

    def root_matches?(cli_root, remote_path, project_root)
      declared = File.join(project_root, remote_path)
      return true if cli_root == declared || cli_root == remote_path

      strip_dot_slash(cli_root) == strip_dot_slash(remote_path)
    end

    def strip_dot_slash(value)
      value.to_s.strip.delete_prefix("./").delete_suffix("/").tr("\\", "/")
    end
  end
end

File: crates/pray-core/src/publish_select.rs
use crate::publish_remote::ManifestPublishRemote;
use crate::{PrayError, PrayResult};
use std::path::{Path, PathBuf};

#[derive(Debug, Clone, PartialEq, Eq)]
pub struct PublishDestination {
    pub name: String,
    pub root: Option<PathBuf>,
    pub server: Option<String>,
    pub packages: Vec<String>,
}

#[derive(Debug, Clone, Default)]
pub struct PublishCliDest {
    pub to: Vec<String>,
    pub roots: Vec<PathBuf>,
    pub servers: Vec<String>,
}

pub fn resolve_publish_destinations(
    remotes: &[ManifestPublishRemote],
    cli: &PublishCliDest,
    project_root: &Path,
) -> PrayResult<Vec<PublishDestination>> {
    if !cli.to.is_empty() && (!cli.roots.is_empty() || !cli.servers.is_empty()) {
        return Err(PrayError::Usage(
            "publish --to cannot be combined with --root or --server".to_string(),
        ));
    }
    if remotes.is_empty() {
        return resolve_undeclared_destinations(cli);
    }
    let selected = select_remotes(remotes, cli, project_root)?;
    Ok(selected
        .into_iter()
        .map(|remote| PublishDestination {
            name: remote.name.clone(),
            root: remote.path.as_ref().map(|path| project_root.join(path)),
            server: remote.url.clone(),
            packages: remote.packages.clone(),
        })
        .collect())
}

pub fn resolve_path_remote(
    remotes: &[ManifestPublishRemote],
    to: Option<&str>,
    root: Option<&Path>,
    project_root: &Path,
) -> PrayResult<PathBuf> {
    let path_remotes: Vec<&ManifestPublishRemote> = remotes
        .iter()
        .filter(|remote| remote.path.is_some())
        .collect();
    if let Some(name) = to {
        let remote = remotes
            .iter()
            .find(|remote| remote.name == name)
            .ok_or_else(|| PrayError::Usage(format!("unknown publish remote: {name}")))?;
        let path = remote.path.as_ref().ok_or_else(|| {
            PrayError::Usage(format!(
                "publish remote {name} is a URL; yank, serve, and token need a path remote"
            ))
        })?;
        return Ok(project_root.join(path));
    }
    if remotes.is_empty() {
        return root
            .map(Path::to_path_buf)
            .ok_or_else(|| PrayError::Usage("requires --root PATH".to_string()));
    }
    if let Some(root) = root {
        let matched = remotes.iter().find(|remote| {
            remote
                .path
                .as_ref()
                .is_some_and(|path| root_matches(root, path, project_root))
        });
        if matched.is_none() {
            return Err(PrayError::Usage(format!(
                "path {} is not a declared publish remote",
                root.display()
            )));
        }
        return Ok(root.to_path_buf());
    }
    if path_remotes.len() == 1 {
        if let Some(path) = &path_remotes[0].path {
            return Ok(project_root.join(path));
        }
    }
    Err(PrayError::Usage(
        "say which path remote with --to NAME or --root PATH".to_string(),
    ))
}

fn resolve_undeclared_destinations(cli: &PublishCliDest) -> PrayResult<Vec<PublishDestination>> {
    if !cli.to.is_empty() {
        return Err(PrayError::Usage(
            "Prayfile has no publish remotes; pass --root PATH or --server URL".to_string(),
        ));
    }
    if cli.roots.is_empty() && cli.servers.is_empty() {
        return Err(PrayError::Unsupported(
            "publish requires at least one --root PATH or --server URL".to_string(),
        ));
    }
    let mut dests = Vec::new();
    for root in &cli.roots {
        dests.push(PublishDestination {
            name: root.display().to_string(),
            root: Some(root.clone()),
            server: None,
            packages: Vec::new(),
        });
    }
    for server in &cli.servers {
        dests.push(PublishDestination {
            name: server.clone(),
            root: None,
            server: Some(server.clone()),
            packages: Vec::new(),
        });
    }
    Ok(dests)
}

fn select_remotes<'a>(
    remotes: &'a [ManifestPublishRemote],
    cli: &PublishCliDest,
    project_root: &Path,
) -> PrayResult<Vec<&'a ManifestPublishRemote>> {
    if !cli.to.is_empty() {
        let mut selected = Vec::new();
        for name in &cli.to {
            let remote = remotes
                .iter()
                .find(|remote| remote.name == *name)
                .ok_or_else(|| PrayError::Usage(format!("unknown publish remote: {name}")))?;
            selected.push(remote);
        }
        return Ok(selected);
    }
    if cli.roots.is_empty() && cli.servers.is_empty() {
        return Ok(remotes.iter().collect());
    }
    let mut selected = Vec::new();
    for root in &cli.roots {
        let remote = remotes
            .iter()
            .find(|remote| {
                remote
                    .path
                    .as_ref()
                    .is_some_and(|path| root_matches(root, path, project_root))
            })
            .ok_or_else(|| {
                PrayError::Usage(format!(
                    "path {} is not a declared publish remote",
                    root.display()
                ))
            })?;
        selected.push(remote);
    }
    for server in &cli.servers {
        let remote = remotes
            .iter()
            .find(|remote| remote.url.as_deref() == Some(server.as_str()))
            .ok_or_else(|| {
                PrayError::Usage(format!("server {server} is not a declared publish remote"))
            })?;
        selected.push(remote);
    }
    Ok(selected)
}

pub(crate) fn root_matches(cli_root: &Path, remote_path: &str, project_root: &Path) -> bool {
    let declared = project_root.join(remote_path);
    if cli_root == declared || cli_root == Path::new(remote_path) {
        return true;
    }
    strip_dot_slash(&cli_root.to_string_lossy()) == strip_dot_slash(remote_path)
}

fn strip_dot_slash(value: &str) -> String {
    value
        .trim()
        .trim_start_matches("./")
        .trim_end_matches('/')
        .replace('\\', "/")
}
